src/Builder/BuilderEncoder.php

```diff
@@ -4,323 +4,101 @@
-                foreach (get_object_vars($this->recursiveEncode($value)) as $subKey => $subValue) {
-                    if (property_exists($result, (string) $subKey)) {
-                        throw new LogicException(sprintf('Duplicate key "%s" in query object', $subKey));
-                    }
+        $encoder = $this->getEncoderFor($value);
 
-                    $result->{$subKey} = $subValue;
-                }
-            } else {
-                if (property_exists($result, (string) $key)) {
-                    throw new LogicException(sprintf('Duplicate key "%s" in query object', $key));
-                }
-
-                $result->{$key} = $this->encodeIfSupported($value);
-            }
+        if (! $encoder || ! $encoder->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
         }
 
-        return $result;
+        return $encoder->encode($value);
     }
 
-    /**
-     * For the $setWindowFields stage output parameter, the optional window parameter is encoded in the same object
-     * of the window operator.
-     *
-     * @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/setWindowFields/
-     */
-    private function encodeOutputWindow(OutputWindow $outputWindow): stdClass
+    private function getEncoderFor(object $value): ExpressionEncoder|null
     {
-        $result = $this->recursiveEncode($outputWindow->operator);
-
-        // Transform the result into an stdClass if a document is provided
-        if (! $outputWindow->operator instanceof WindowInterface && (is_array($result) || is_object($result))) {
-            if (! is_first_key_operator($result)) {
-                throw new LogicException(sprintf('Expected OutputWindow::$operator to be an operator. Got "%s"', array_key_first((array) $result)));
-            }
-
-            $result = (object) $result;
+        $valueClass = $value::class;
+        if (isset($this->cachedEncoders[$valueClass])) {
+            return $this->cachedEncoders[$valueClass];
         }
 
-        if (! $result instanceof stdClass) {
-            throw new LogicException(sprintf('Expected OutputWindow::$operator to be an stdClass, array or WindowInterface. Got "%s"', get_debug_type($result)));
-        }
+        $encoderList = $this->customEncoders + $this->defaultEncoders;
 
-        if ($outputWindow->window !== Optional::Undefined) {
-            $result->window = $this->recursiveEncode($outputWindow->window);
-        }
-
-        return $result;
-    }
-
-    /**
-     * Nested arrays and objects must be encoded recursively.
-     */
-    private function recursiveEncode(mixed $value): mixed
-    {
-        if (is_array($value)) {
-            foreach ($value as $key => $val) {
-                $value[$key] = $this->recursiveEncode($val);
+        // First attempt: match class name exactly
+        foreach ($encoderList as $className => $encoderClass) {
+            if ($className == $valueClass) {
```

Comment:
Any reason not to use `===` here?

Reply:
A missing phpcs rule is the reason. I've enabled the rule and fixed this occurrence.